### PR TITLE
Postgres support

### PR DESCRIPTION
**Pull request by alek-talabat**
# Description

Added postgres support to [kafkaflow-retry-extensions](https://github.com/Farfetch/kafkaflow-retry-extensions) based on the implementation for sql server. It's needed to enable using the package with postgres db. Decided to make it as similar as possible to sql code, to make sure it's aligned to the existing style & practices and does not cause any discrepancies. Further adjustments of extracting shared code could be done in separate PRs, after confirming it with the repo owners. All of the postgres adjustments (comparing to sql code/scripts) can be viewed by [skipping first commit](https://github.com/Farfetch/kafkaflow-retry-extensions/pull/108/files/0ac38a3e9e3857b6823d8f3c8282e681383b4953..f28ad9f820f5af0c19e772b1ad8edf7413e34dea#diff-086cd55f111db6ff0b0fefbd6da708cb04b09d0377ac972b6c21038dc5b4896c) 

## How Has This Been Tested?

Run all unit tests and integration tests locally to make sure that new integration works. New integration is covered same way as sql server

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have added tests to cover my changes
-   [x] I have made corresponding changes to the documentation

### Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

**Comment by alek-talabat** [comment]
Would appreciate Your review @luispfgarces @filipeesch @fernando-a-marins

**Review by luispfgarces — COMMENTED** [review]
Implementation seems to be very alike the SQL Server implementation. Many of the things are duplicated, but I agree that this could not be accomplished without doing it (at least without a structural refactor).

I am leaving just 2 comments on minor things to solve, nothing that poses a big deal. Thank you @alek-talabat for your contribution. On my end, I'll approve once you fix those 👌

**Comment by luispfgarces** [comment]
We need to merge #93 to fix the build pipeline (support for builds with forked repositories).

File: src/KafkaFlow.Retry.IntegrationTests/Core/Bootstrappers/BootstrapperKafka.cs [hunk 1]
```diff
@@ -1,4 +1,6 @@
-﻿namespace KafkaFlow.Retry.IntegrationTests.Core.Bootstrappers
+﻿using KafkaFlow.Retry.Postgres;
```
[the remark this answers is not in the document]
Comment (alek-talabat) [reply src/KafkaFlow.Retry.IntegrationTests/Core/Bootstrappers/BootstrapperKafka.cs]: Sure, all usings were moved inside namespace

File: src/KafkaFlow.Retry.Postgres/Repositories/RetryQueueItemRepository.cs [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+            }
+        }
+
+        private async Task<IList<RetryQueueItemDbo>> ExecuteReaderAsync(NpgsqlCommand command)
+        {
+            var items = new List<RetryQueueItemDbo>();
+
+            using (var reader = await command.ExecuteReaderAsync().ConfigureAwait(false))
+            {
+                while (await reader.ReadAsync())
```
[the remark this answers is not in the document]
Comment (alek-talabat) [reply src/KafkaFlow.Retry.Postgres/Repositories/RetryQueueItemRepository.cs]: Added missing  `ConfigureAwait(false)`

**Review by luispfgarces — DISMISSED** [review]
Thank you for your changes @alek-talabat. LGTM

**Comment by luispfgarces** [comment]
@alek-talabat the PR #93 was merged, which will fix your build pipeline. Please merge from main branch.

**Comment by martinhonovais** [comment]
Hi @alek-talabat,
First of all, many thanks for your contribution, we are very glad to review it. 
Everything seemed good to me but I have some considerations:
1. I feel that we could have a sample using the PostgresSql repo. This could help the PostgresSql users experimenting the repo with not too much effort.
2. Last week we merged to main a [feat ](https://github.com/Farfetch/kafkaflow-retry-extensions/commit/86daad24f26ebbd86de68e01d50fbd5d965c415c) that allows the dev to apply a retention policy for queues that have been done. Please take it into consideration in your PR as well. It will help devs to save repo space and money.
3. Let us know if you did some API's tests using PostgreSql as a repo and if it went well.

**Comment by luispfgarces** [comment]
@alek-talabat can you please check the build error?

> Error: /home/runner/work/kafkaflow-retry-extensions/kafkaflow-retry-extensions/src/KafkaFlow.Retry.Postgres/RetryQueueDataProvider.cs(18,52): error CS0535: 'RetryQueueDataProvider' does not implement interface member 'IRetryDurableQueueRepositoryProvider.DeleteQueuesAsync(DeleteQueuesInput)' [/home/runner/work/kafkaflow-retry-extensions/kafkaflow-retry-extensions/src/KafkaFlow.Retry.Postgres/KafkaFlow.Retry.Postgres.csproj]

I think that the pull request that went Live before yours created a new method on the repository interface, which yours does not have (because you implemented it before this change).

Thank you! 🙏

**Comment by alek-talabat** [comment]
Sorry for the delay, but changes from the most recent master are already included. Would appreciate Your review @martinhonovais @gsferreira to have this one closed

**Review by carlosgoias — APPROVED** [review]

**Review by gsferreira — APPROVED** [review]

**Pull request merged**